### PR TITLE
Use dhcp leases contained in net-dumpxml to get VM IP

### PR DESCRIPTION
**Pull request by guillaumerose**
Move away from reading the IP from the dnsmasq file, and instead use
the equivalent of virsh net-dumpxml command.
The driver doesnt assume anything about the filesystem and only uses the
libvirt API.

### CODE REVIEW

**Comment by gbraad** [comment]
Please  provide a more clear description of the changes. In short it should be "move away from reading the IP from the leases file, and instead use the virsh net-dumpxml command". right?

**Comment by guillaumerose** [comment]
done :+1:

**Review by cfergeau — COMMENTED** [review]
Maybe https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainInterfaceAddresses could be used rather than doing our own parsing?

**Comment by guillaumerose** [comment]
Good idea! done

File: pkg/libvirt/libvirt.go [hunk 1]
```diff
@@ -651,12 +574,21 @@ func (d *Driver) GetIP() (string, error) {
...
 	}
-	mac, err := d.getMAC()
+	ifaces, err := d.vm.ListAllInterfaceAddresses(libvirt.DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE)
 	if err != nil {
 		return "", err
 	}
-
-	return d.getIPByMacFromSettings(mac)
+	if len(ifaces) != 1 {
+		log.Errorf("unexpected number of interfaces for domain %s", d.MachineName)
```
Comment (cfergeau) [inline pkg/libvirt/libvirt.go]: Or the list could be filtered by mac address (`const macAddress = "52:fd:fc:07:21:82"`), and then we'd error out if there is more than one matching interface?

File: pkg/libvirt/libvirt.go [hunk 2]
```diff
@@ -651,12 +574,21 @@ func (d *Driver) GetIP() (string, error) {
...
+		return "", nil
+	}
+	if len(ifaces[0].Addrs) != 1 {
+		log.Errorf("unexpected number of IPs for domain %s", d.MachineName)
```
Comment (cfergeau) [inline pkg/libvirt/libvirt.go]: I'd also do some filtering here as maybe on some setups we'll be getting both ipv4 and ipv6 addresses? Don't error out, just return the first ipv4 address, with a warning that there were multiple addresses set on the interface?

**Comment by guillaumerose** [comment]
all done

File: pkg/libvirt/libvirt.go [hunk 3]
```diff
@@ -641,12 +564,21 @@ func (d *Driver) GetIP() (string, error) {
...
+	ifaces, err := d.vm.ListAllInterfaceAddresses(libvirt.DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE)
 	if err != nil {
 		return "", err
 	}
-
-	return d.getIPByMacFromSettings(mac)
+	for _, iface := range ifaces {
+		if iface.Hwaddr == macAddress {
+			for _, addr := range iface.Addrs {
+				if addr.Type == 0 { // ipv4
```
Comment (cfergeau) [inline pkg/libvirt/libvirt.go]: `addr.Type == int(libvirt.IP_ADDR_TYPE_IPV4)`

**Review by cfergeau — APPROVED** [review]
One minor comment left, I'll check with the libvirt people if it's intentional that `DomainIPAddress.Type` is not using `libvirt.IPAddrType`.

**Pull request merged**